# Lab 9 프로그래밍 과제

In [64]:
import numpy as np

# 1. NumPy의 np.dot()함수를 이용하여 주어진 두 개의 vector로 dot product를 수행함.

In [65]:
vector_x = np.arange(4)
vector_y = np.arange(4)
vector_x, vector_y

(array([0, 1, 2, 3]), array([0, 1, 2, 3]))

In [66]:
vec_vec_mul_result = np.dot(vector_x, vector_y)
print(vec_vec_mul_result)

14


NumPy의 vector_x, vector_y 벡터를 1차원 리스트로 변환

In [67]:
list_1D_x = vector_x.tolist()
list_1D_y = vector_y.tolist()
list_1D_x, list_1D_y

([0, 1, 2, 3], [0, 1, 2, 3])

# Problem 1 (2 points): 두 개의 리스트 list_1D_x와 list_1D_y를 이용하여 np.dot(vector_x, vector_y) 함수와 같은 계산을 수행하는 vec_vec_mul_func() 함수를 작성하시오.

(참고: NumPy에서 제공하는 함수를 (ex, np.sum(), np.dot() 등) 이용하지 말고 반복문을 이용해서 코드를 작성할 것.)

In [68]:
"""
Problem 1: 이 셀 안에 코드를 작성하시오.
"""
# 함수 작성
def vec_vec_mul_func(vec_x, vec_y):
  result_vec=0
  for i in range(len(vec_x)):
    result_vec += vec_x[i]*vec_y[i]

  return result_vec

In [69]:
# 함수 호출
list_vec_vec_mul_result = vec_vec_mul_func(list_1D_x, list_1D_y)

# 결과 출력
print(list_vec_vec_mul_result)

14


# 2. NumPy의 np.dot()함수를 이용하여 주어진 하나의 matrix와 하나의 vector로 matrix-vector multiplication을 수행함.

In [70]:
matrix_A = np.arange(32).reshape(8,4)
vector_x = np.arange(4)
matrix_A, vector_x

(array([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23],
        [24, 25, 26, 27],
        [28, 29, 30, 31]]),
 array([0, 1, 2, 3]))

In [71]:
mat_vec_mul_result = np.dot(matrix_A, vector_x)
print(mat_vec_mul_result)

[ 14  38  62  86 110 134 158 182]


NumPy의 matrix_A 행렬을 2차원 리스트로, 그리고 vector_x 벡터를 1차원 리스트로 변환

In [76]:
list_2D_A = matrix_A.tolist()
list_1D_x = vector_x.tolist()
print(len(list_2D_A))

8


# Problem 2 (3 points): 두 개의 리스트 list_2D_A와 list_1D_x를 이용하여 np.dot(matrix_A, vector_x) 함수와 같은 계산을 수행하는 mat_vec_mul_func() 함수를 작성하시오.

(참고: NumPy에서 제공하는 함수를 (ex, np.sum(), np.dot() 등) 이용하지 말고 반복문을 이용해서 코드를 작성할 것.)

In [117]:
"""
Problem 2: 이 셀 안에 코드를 작성하시오.
"""
# 함수 작성
def mat_vec_mul_func(mat_A, vec_x):
  result_vec=[0 for i in range(len(mat_A))]
  for i in range(len(mat_A)):
    for j in range(len(vec_x)):
      result_vec[i] += mat_A[i][j]*vec_x[j]

  return result_vec

In [118]:
# 함수 호출
list_mat_vec_mul_result = mat_vec_mul_func(list_2D_A, list_1D_x)

# 결과 출력
print(list_mat_vec_mul_result)

[14, 38, 62, 86, 110, 134, 158, 182]


# 3. NumPy의 np.dot()함수를 이용하여 주어진 두 개의 matrix로 matrix-matrix multiplication을 수행함.

In [79]:
matrix_A = np.arange(32).reshape(8,4)
matrix_B = np.arange(64).reshape(4,16)
matrix_A, matrix_B

(array([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23],
        [24, 25, 26, 27],
        [28, 29, 30, 31]]),
 array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31],
        [32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47],
        [48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]]))

In [80]:
mat_mat_mul_result = np.dot(matrix_A, matrix_B)
print(mat_mat_mul_result)

[[ 224  230  236  242  248  254  260  266  272  278  284  290  296  302
   308  314]
 [ 608  630  652  674  696  718  740  762  784  806  828  850  872  894
   916  938]
 [ 992 1030 1068 1106 1144 1182 1220 1258 1296 1334 1372 1410 1448 1486
  1524 1562]
 [1376 1430 1484 1538 1592 1646 1700 1754 1808 1862 1916 1970 2024 2078
  2132 2186]
 [1760 1830 1900 1970 2040 2110 2180 2250 2320 2390 2460 2530 2600 2670
  2740 2810]
 [2144 2230 2316 2402 2488 2574 2660 2746 2832 2918 3004 3090 3176 3262
  3348 3434]
 [2528 2630 2732 2834 2936 3038 3140 3242 3344 3446 3548 3650 3752 3854
  3956 4058]
 [2912 3030 3148 3266 3384 3502 3620 3738 3856 3974 4092 4210 4328 4446
  4564 4682]]


NumPy의 matrix_A 와 matrix_B 행렬을 2차원 리스트로 변환

In [98]:
list_2D_A = matrix_A.tolist()
list_2D_B = matrix_B.tolist()
list_2D_A, list_2D_B
print(len(list_2D_A))

8


# Problem 3 (4 points): 두 개의 리스트 list_2D_A와 list_2D_B를 이용하여 np.dot(matrix_A, matrix_B) 함수와 같은 계산을 수행하는 mat_mat_mul_func() 함수를 작성하시오.

(참고: NumPy에서 제공하는 함수를 (ex, np.sum(), np.dot() 등) 이용하지 말고 반복문을 이용해서 코드를 작성할 것.)

In [113]:
"""
Problem 3: 이 셀 안에 코드를 작성하시오.
"""
# 함수 작성
def mat_mat_mul_func(mat_A, mat_B):
  result_mat=[[0 for i in range(matrix_B.shape[1])] for i in range(len(mat_A))]
  for i in range(len(mat_A)):
    for j in range(matrix_B.shape[1]):
      sum=0
      for k in range(len(mat_B)):
        sum += mat_A[i][k]*mat_B[k][j]
      result_mat[i][j] += sum


  return result_mat

위에 구현한 row_sum() 함수를 호출하여 반환된 결과값을 출력함.

In [114]:
# 함수 호출
list_mat_mat_mul_result = mat_mat_mul_func(list_2D_A, list_2D_B)

# 결과 출력
print(list_mat_mat_mul_result)

[[224, 230, 236, 242, 248, 254, 260, 266, 272, 278, 284, 290, 296, 302, 308, 314], [608, 630, 652, 674, 696, 718, 740, 762, 784, 806, 828, 850, 872, 894, 916, 938], [992, 1030, 1068, 1106, 1144, 1182, 1220, 1258, 1296, 1334, 1372, 1410, 1448, 1486, 1524, 1562], [1376, 1430, 1484, 1538, 1592, 1646, 1700, 1754, 1808, 1862, 1916, 1970, 2024, 2078, 2132, 2186], [1760, 1830, 1900, 1970, 2040, 2110, 2180, 2250, 2320, 2390, 2460, 2530, 2600, 2670, 2740, 2810], [2144, 2230, 2316, 2402, 2488, 2574, 2660, 2746, 2832, 2918, 3004, 3090, 3176, 3262, 3348, 3434], [2528, 2630, 2732, 2834, 2936, 3038, 3140, 3242, 3344, 3446, 3548, 3650, 3752, 3854, 3956, 4058], [2912, 3030, 3148, 3266, 3384, 3502, 3620, 3738, 3856, 3974, 4092, 4210, 4328, 4446, 4564, 4682]]


# Problem 4 (1 point)
## NumPy에서 제공되는 np.dot() 함수와 NumPy배열에서 변환된 list를 이용하여 직접 구현한 함수 vec_vec_mul_func(), mat_vec_mul_func(), mat_mat_mul_func() 의 실행 속도를 측정하여 비교하시오.
## 실행시간을 비교한 결과를 바탕으로 결론을 짧게 요약해서 작성하시오.
### 위 문제들에서 주어진 Vector 와 Matrix 의 크기보다 더 큰 사이즈의 Vector 와 Matrix를 이용해서 실험해보는 것을 권장함.
### (예, 사이즈가 10000x10000 인 Matrix, 사이즈가 10000 인 Vector)
#### (Markdown 과 Code를 활용하여 자유형식으로 아래에 결론을 작성 (셀 추가 가능) )

In [94]:
# 실행 속도를 측정하는 코드 예
import time

start = time.time()

# 실행 속도를 측정할 코드
print("hello")

end = time.time()

print(end - start)

hello
0.0007045269012451172


In [119]:
"""
Problem 4: 실행 속도를 비교하고 및 결론을 작성하시오. (아래 셀 추가 가능)
"""
matrix_A = np.arange(40000).reshape(200,200)
matrix_D = np.arange(40000).reshape(200,200)
vector_B = np.arange(200)
vector_C = np.arange(200)

start=time.time()
np.dot(matrix_A,vector_B)
end=time.time()
print("np.dot:",end-start)

start=time.time()
np.dot(vector_C,vector_B)
end=time.time()
print("np.dot:",end-start)

start=time.time()
np.dot(matrix_A,matrix_D)
end=time.time()
print("np.dot:",end-start)

matrix_a=matrix_A.tolist()
vector_b=vector_B.tolist()
vector_c=vector_C.tolist()
matrix_d=matrix_D.tolist()

start=time.time()
mat_vec_mul_func(matrix_a,vector_b)
end=time.time()
print("mat_vec_mul_func():",end-start)

start=time.time()
vec_vec_mul_func(vector_c,vector_b)
end=time.time()
print("vec_vec_mul_func():",end-start)

start=time.time()
mat_mat_mul_func(matrix_a,matrix_d)
end=time.time()
print("mat_mat_mul_func():",end-start)






np.dot: 0.0002474784851074219
np.dot: 0.0001456737518310547
np.dot: 0.011247873306274414
mat_vec_mul_func(): 0.00722503662109375
vec_vec_mul_func(): 9.1552734375e-05
mat_mat_mul_func(): 0.10848498344421387


#Numpy에서 제공한 np.dot()함수의 계산 속도가 직접 만든 함수의 속도보다 빠르다.
##np.dot: 0.0002474784851074219 vs mat_vec_mul_func(): 0.00722503662109375
##np.dot: 0.0001456737518310547 vs vec_vec_mul_func(): 9.1552734375e-05
##np.dot: 0.011247873306274414 vs mat_mat_mul_func(): 0.10848498344421387